### PR TITLE
[GenericPoseGraph] Add PoseGraphController.

### PR DESCRIPTION
**Pull request by pifon2a**

### CODE REVIEW

File: cartographer/pose_graph/pose_graph_data.cc [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+namespace {
+
+std::unique_ptr<Constraint> CreateConstraint(
+    const proto::Constraint& constraint) {
+  const auto& id = constraint.id();
+  const auto& loss = constraint.loss_function();
+  const auto& cost = constraint.cost_function();
+  switch (cost.type_case()) {
+    case (proto::CostFunction::kRelativePose2D):
+      return common::make_unique<RelativePoseConstraint2D>(
```
Comment (mbokeloh) [inline cartographer/pose_graph/pose_graph_data.cc]: I like this pattern much better than the one used in the nodes. What do you think about having a CreateNode() function as well?

File: cartographer/pose_graph/pose_graph_data.cc [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+      return;
+    }
+    case (proto::Parameters::TYPE_NOT_SET): {
+      LOG(FATAL) << "Node parameter type is not set.";
+      return;
+    }
+  }
+}
+
+void AddConstraintToPoseGraphData(const proto::Constraint& constraint,
```
Comment (mbokeloh) [inline cartographer/pose_graph/pose_graph_data.cc]: How about this:
Expose the CreateConstraint/CreateNode functions instead and fill the vector/map in the caller. I think this is more flexible than the current interface.

File: cartographer/pose_graph/pose_graph_data.cc (hunk 1 above)
Comment (pifon2a) [reply cartographer/pose_graph/pose_graph_data.cc]: How would it look like? Nodes are not really stored by keeping a ptr to the base class.

File: cartographer/pose_graph/pose_graph_data.cc (hunk 2 above)
Comment (pifon2a) [reply cartographer/pose_graph/pose_graph_data.cc]: I am not sure that I understand how to do that with nodes.

File: cartographer/pose_graph/node/imu_calibration.cc [hunk 3]
```diff
@@ -20,12 +20,13 @@ namespace cartographer {
...
 ImuCalibration::ImuCalibration(const NodeId& node_id, bool constant,
-                               double gravity_constant,
-                               const Eigen::Quaterniond& orientation)
+                               const proto::ImuCalibration& imu_calibration)
     : Node(node_id, constant),
-      gravity_constant_(gravity_constant),
-      orientation_{{orientation.x(), orientation.y(), orientation.z(),
-                    orientation.w()}} {}
+      gravity_constant_(imu_calibration.gravity_constant()),
+      orientation_{{imu_calibration.orientation().x(),
```
[the remark this answers is not in the document]
Comment (pifon2a) [reply cartographer/pose_graph/node/imu_calibration.cc]: We definitely need that. Not in this CL.

File: cartographer/pose_graph/pose_graph_data.cc (hunk 1 above)
Comment (mbokeloh) [reply cartographer/pose_graph/pose_graph_data.cc]: Oh right, my bad. Carry on :-)

File: cartographer/pose_graph/pose_graph_data.cc (hunk 2 above)
Comment (mbokeloh) [reply cartographer/pose_graph/pose_graph_data.cc]: Yes, let's keep as is.

**Comment by pifon2a** [comment]
@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user pifon2a. Merge queue now has a length of 1.

**Comment by wally-the-cartographer** [comment]
Error: exit status 1

Log:
~~~
=> handle work item: PR 1335 on googlecartographer/cartographer.
 => Switching to branch master.
 $ git reset --hard HEAD [data/cartographer]
HEAD is now at 5b44305e [ABSL] Purge common::make_unique. (#1340)
 $ git clean -fd [data/cartographer]
 $ git checkout master [data/cartographer]
Already on 'master'
Your branch is up to date with 'origin/master'.
 $ git pull [data/cartographer]
Your configuration specifies to merge with the ref 'refs/heads/master'
from the remote, but no such ref was fetched.

~~~

**Pull request merged**